hoomd/mpcd/SphereGeometry.h

```diff
@@ -0,0 +1,195 @@
+             * when a particle is reset to the boundary location. A particle landing exactly on the boundary from the bulk
+             * can be immediately reflected on the next streaming step, and so the motion is essentially equivalent up to
+             * an epsilon of difference in the channel width.
+             */
+            double r2 = pos.x*pos.x + pos.y*pos.y + pos.z*pos.z;
+            const char sign = (r2 > R*R);
+            // exit immediately if no collision is found or particle is not moving normal (radial direction) to the wall
+            // (since no new collision could have occurred if there is no normal motion)
+            /*
+            r(t) = sqrt(x**2 + y**2 + z**2)
+            radial velocity, r_dot = (x*x_dot + y*y_dot + z*z_dot)/r(t)
+            */
+            float vr = (pos.x*vel.x + pos.y*vel.y + pos.z*vel.z)/sqrt(r2);
+            if (sign == 0 || vr == Scalar(0))
+               {
+               dt = Scalar(0);
+               return false;
+               }
+
+            // Find the point of contact when the particle is just leaving the surface
+            /*
+            * Calculating the time spent outside-bounds requires the knowledge of the point of contact on the boundary.
+            * The point of contact can be calculated using geometrical considerations.
+            *
+            * We know the following quantities,
+            *    1. r(t+del_t)      (the point outside the sphere)
+            *    2. v(t)            (previous velocity)
+            *
+            * Assuming 'r*' is the point of contact on the spherical shell and dx* is the distance travelled by the
+            * particle outside the boundary.
+            *
+            * r* = r(t+del_t) + dx* * v(t)/|v|          (the -ve signs cancel out in the 2nd term)
+            * and |r|**2 = R**2
+            *
+            * Solving the above equations,
+            * dx* = -r(t+del_t).v(t)/|v| +- sqrt((r(t+del_t).v(t))**2 - |r(t+del_t)|**2 + R**2)
+            * consequently, dt = dx* / |v|
+            */
+            double mod_v,r_dot_v_hat,dx_;
+
+            mod_v = sqrt(vel.x*vel.x + vel.y*vel.y + vel.z*vel.z);
+            r_dot_v_hat = (pos.x*vel.x + pos.y*vel.y + pos.z*vel.z)/mod_v;
+
+            dx_ = -r_dot_v_hat + sqrt(r_dot_v_hat*r_dot_v_hat - r2 + R*R);
+            dt = dx_/mod_v;
+
+            // backtrack the particle for time dt to get to point of contact
+            pos.x -= vel.x*dt;
+            pos.y -= vel.y*dt;
+            pos.z -= vel.z*dt;
+
+            // update velocity according to boundary conditions
+            // no-slip requires reflection of the tangential components
+            if (m_bc == boundary::no_slip)
+                {
+                /*
+                Both the tangential and perpendicular components must be reflected.
+                */
+                vel.x = -vel.x;
+                vel.y = -vel.y;
+                vel.z = -vel.z;
+                }
```

Comment:
You need to split the velocity into normal and tangent components here based on the point of intersection (`pos` after the update). Then, you always flip the normal component and optionally flip the tangent component if the boundary is no slip. Arash can help you with this math.

Reply:
The code for the slip boundary conditions is still missing.